File: scripts/instagram_scraper.py
#!/usr/bin/env python3
"""
Instagram Profile Scraper
Kullanim: python instagram_scraper.py <username>
"""

import sys
import json
import time
import random
import os
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.chrome.service import Service

class InstagramScraper:
    def __init__(self):
        self.driver = None

    def setup_driver(self):
        """Chrome driver'ini bot detection bypass ile kur"""
        try:
            options = Options()
            options.add_argument('--headless')
            options.add_argument('--no-sandbox')
            options.add_argument('--disable-dev-shm-usage')
            options.add_argument('--disable-gpu')
            options.add_argument('--window-size=1920,1080')
            options.add_argument('--disable-blink-features=AutomationControlled')
            options.add_experimental_option("excludeSwitches", ["enable-automation"])
            options.add_experimental_option('useAutomationExtension', False)

            user_agents = [
                'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
                'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
                'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
            ]
            options.add_argument(f'--user-agent={random.choice(user_agents)}')

            chrome_driver_path = os.path.join(os.getcwd(), 'node_modules', 'chromedriver', 'lib', 'chromedriver', 'chromedriver.exe')

            if os.path.exists(chrome_driver_path):
                service = Service(chrome_driver_path)
                self.driver = webdriver.Chrome(service=service, options=options)
            else:
                self.driver = webdriver.Chrome(options=options)

            self.driver.execute_script("Object.defineProperty(navigator, 'webdriver', {get: () => undefined})")
            return True

        except Exception as e:
            sys.stderr.write(f"Driver setup error: {e}\n")
            return False

    def scrape_profile(self, username):
        """Instagram profilini scrape et"""
        try:
            url = f"https://www.instagram.com/{username.replace('@', '')}/"
            sys.stderr.write(f"Scraping: {url}\n")

            self.driver.get(url)
            time.sleep(random.uniform(8, 12))

            page_source = self.driver.page_source
            sys.stderr.write(f"Page source length: {len(page_source)}\n")

            title = self.driver.title
            sys.stderr.write(f"Page title: {title}\n")

            current_url = self.driver.current_url
            sys.stderr.write(f"Current URL: {current_url}\n")

            if 'login' in current_url.lower() or 'accounts/login' in current_url:
                sys.stderr.write("Redirected to login page - Instagram blocking\n")
                return None

            profile_data = {
                'username': username,
                'followers': self.get_followers(),
                'following': self.get_following(),
                'posts': self.get_posts_count(),
                'bio': self.get_bio(),
                'is_verified': self.is_verified(),
                'is_private': self.is_private()
            }

            sys.stderr.write(f"Scraped data: {profile_data}\n")
            return profile_data

        except Exception as e:
            sys.stderr.write(f"Scraping error: {e}\n")
            return None

    def get_followers(self):
        selectors = [
            'a[href*="/followers/"] span',
            'header section ul li:nth-child(2) span',
            'span[title]'
        ]

        for selector in selectors:
            try:
                elements = self.driver.find_elements(By.CSS_SELECTOR, selector)
                for element in elements:
                    text = element.get_attribute('title') or element.text
                    if text and ('K' in text or 'M' in text or text.replace(',', '').isdigit()):
                        return self.parse_count(text)
            except:
                continue

        return 0

    def get_following(self):
        selectors = [
            'a[href*="/following/"] span',
            'header section ul li:nth-child(3) span'
        ]

        for selector in selectors:
            try:
                elements = self.driver.find_elements(By.CSS_SELECTOR, selector)
                for element in elements:
                    text = element.text
                    if text and ('K' in text or 'M' in text or text.replace(',', '').isdigit()):
                        return self.parse_count(text)
            except:
                continue

        return 0

    def get_posts_count(self):
        selectors = [
            'header section ul li:first-child span',
            'div[data-testid="user-posts-count"]'
        ]

        for selector in selectors:
            try:
                elements = self.driver.find_elements(By.CSS_SELECTOR, selector)
                for element in elements:
                    text = element.text
                    if text and ('K' in text or 'M' in text or text.replace(',', '').isdigit()):
                        return self.parse_count(text)
            except:
                continue

        return 0

    def get_bio(self):
        selectors = [
            'header section div div span',
            'div[data-testid="user-bio"]',
            'header section div span',
            'article header section div div span'
        ]

        for selector in selectors:
            try:
                elements = self.driver.find_elements(By.CSS_SELECTOR, selector)
                for element in elements:
                    text = element.text.strip()
                    if (
                        text and len(text) > 10 and
                        'posts' not in text.lower() and
                        'followers' not in text.lower() and
                        'following' not in text.lower()
                    ):
                        return text
            except:
                continue

        return ""

    def is_verified(self):
        try:
            self.driver.find_element(By.CSS_SELECTOR, '[aria-label*="Verified"]')
            return True
        except:
            return False

    def is_private(self):
        try:
            page_source = self.driver.page_source
            return 'This account is private' in page_source
        except:
            return False

    def parse_count(self, text):
        if not text:
            return 0

        clean_text = text.replace(',', '').replace('.', '').lower()

        try:
            if 'k' in clean_text:
                return int(float(clean_text.replace('k', '')) * 1000)
            elif 'm' in clean_text:
                return int(float(clean_text.replace('m', '')) * 1000000)
            else:
                return int(clean_text)
        except:
            return 0

    def close(self):
        if self.driver:
            self.driver.quit()


def main():
    if len(sys.argv) != 2:
        sys.stderr.write("Usage: python instagram_scraper.py <username>\n")
        sys.exit(1)

    username = sys.argv[1]
    scraper = InstagramScraper()

    try:
        if not scraper.setup_driver():
            print(json.dumps({"error": "Failed to setup driver"}))
            sys.exit(1)

        profile_data = scraper.scrape_profile(username)

        if profile_data:
            print(json.dumps(profile_data))
        else:
            print(json.dumps({"error": "Failed to scrape profile"}))

    except Exception as e:
        print(json.dumps({"error": str(e)}))

    finally:
        scraper.close()


if __name__ == "__main__":
    main()

File: lib/scraping/instagram-scraper.ts
import { Builder, By, until, WebDriver } from 'selenium-webdriver';
import { Options as ChromeOptions, ServiceBuilder } from 'selenium-webdriver/chrome';

interface InstagramProfile {
  username: string;
  followers: number;
  following: number;
  posts: number;
  bio: string;
  isVerified: boolean;
  isPrivate: boolean;
  recentPosts: InstagramPost[];
}

interface InstagramPost {
  imageUrl: string;
  caption: string;
  likes: number;
  comments: number;
  timestamp: string;
  hashtags: string[];
}

export class InstagramScraper {
  private driver: WebDriver | null = null;

  async initialize() {
    const options = new ChromeOptions();
    options.addArguments('--headless'); // Headless mode geri aç
    options.addArguments('--no-sandbox');
    options.addArguments('--disable-dev-shm-usage');
    options.addArguments('--disable-gpu');
    options.addArguments('--window-size=1920,1080');
    
    // Advanced bot detection bypass
    options.addArguments('--disable-blink-features=AutomationControlled');
    options.addArguments('--disable-web-security');
    options.addArguments('--disable-features=VizDisplayCompositor');
    options.addArguments('--disable-extensions');
    options.addArguments('--disable-plugins');
    options.addArguments('--disable-images'); // Hızlandırmak için
    options.addArguments('--disable-javascript'); // JS'i devre dışı bırak
    options.addArguments('--disable-css'); // CSS'i devre dışı bırak
    
    // Realistic user agent
    const userAgents = [
      'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
      'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36',
      'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
    ];
    const randomUA = userAgents[Math.floor(Math.random() * userAgents.length)];
    options.addArguments(`--user-agent=${randomUA}`);
    
    // Bot detection bypass
    options.excludeSwitches(['enable-automation']);
    options.addArguments('--disable-blink-features=AutomationControlled');
    
    // Proxy support (opsiyonel - şimdilik kapalı)
    // options.addArguments('--proxy-server=http://proxy-server:port');
    
    // Chrome driver path'ini belirt
    const path = require('path');
    const chromedriverPath = path.join(process.cwd(), 'node_modules', 'chromedriver', 'lib', 'chromedriver', 'chromedriver.exe');
    
    this.driver = await new Builder()
      .forBrowser('chrome')
      .setChromeOptions(options)
      .setChromeService(new ServiceBuilder(chromedriverPath))
      .build();
  }

  async scrapeProfile(username: string): Promise<InstagramProfile | null> {
    if (!this.driver) {
      throw new Error('Scraper not initialized');
    }

    try {
      // Instagram profiline git
      const profileUrl = `https://www.instagram.com/${username.replace('@', '')}/`;
      console.log(`Scraping Instagram profile: ${profileUrl}`);
      await this.driver.get(profileUrl);

      // Sayfanın yüklenmesini bekle - birden fazla selector dene
      console.log('Waiting for page to load...');
      
      const pageSelectors = ['article', 'main', 'section', 'div[role="main"]', 'header'];
      let pageLoaded = false;
      
      for (const selector of pageSelectors) {
        try {
          await this.driver.wait(until.elementLocated(By.css(selector)), 5000);
          console.log(`Page loaded with selector: ${selector}`);
          pageLoaded = true;
          break;
        } catch (e) {
          console.log(`Selector ${selector} not found, trying next...`);
          continue;
        }
      }
      
      if (!pageLoaded) {
        console.log('Page load timeout, but continuing...');
      }
      
      // Biraz daha bekle - sayfa tamamen yüklensin
      await new Promise(resolve => setTimeout(resolve, 5000));
      
      console.log('Extracting profile data...');

      // Profil bilgilerini çek
      const profile: InstagramProfile = {
        username,
        followers: await this.getFollowerCount(),
        following: await this.getFollowingCount(),
        posts: await this.getPostCount(),
        bio: await this.getBio(),
        isVerified: await this.isVerified(),
        isPrivate: await this.isPrivate(),
        recentPosts: await this.getRecentPosts()
      };

      return profile;
    } catch (error) {
      console.error(`Error scraping ${username}:`, error);
      return null;
    }
  }

  private async getFollowerCount(): Promise<number> {
    try {
      // Birden fazla selector dene
      const selectors = [
        "//a[contains(@href, '/followers/')]/span",
        "//a[contains(@href, '/followers/')]/span/span",
        "//span[contains(text(), 'followers')]/../span[1]",
        "//div[contains(text(), 'followers')]",
        "//span[contains(text(), 'followers')]"
      ];

      for (const selector of selectors) {
        try {
          const element = await this.driver!.findElement(By.xpath(selector));
          const text = await element.getText();
          console.log(`Follower selector: ${selector}, text: ${text}`);
          
          if (text && (text.includes('K') || text.includes('M') || /^\d+$/.test(text.replace(/,/g, '')))) {
            return this.parseCount(text);
          }
        } catch (e) {
          continue;
        }
      }

      // CSS selectors da dene
      const cssSelectors = [
        'a[href*="/followers/"] span',
        'header section ul li:nth-child(2) span',
        'header section ul li span'
      ];

      for (const selector of cssSelectors) {
        try {
          const elements = await this.driver!.findElements(By.css(selector));
          for (const element of elements) {
            const text = await element.getText();
            if (text && (text.includes('K') || text.includes('M') || /^\d+$/.test(text.replace(/,/g, '')))) {
              console.log(`Follower CSS selector: ${selector}, text: ${text}`);
              return this.parseCount(text);
            }
          }
        } catch (e) {
          continue;
        }
      }

      console.log('Follower count not found');
      return 0;
    } catch (error) {
      console.error('Error getting follower count:', error);
      return 0;
    }
  }

  private async getFollowingCount(): Promise<number> {
    try {
      const element = await this.driver!.findElement(
        By.xpath("//a[contains(@href, '/following/')]/span")
      );
      const text = await element.getText();
      return this.parseCount(text);
    } catch {
      return 0;
    }
  }

  private async getPostCount(): Promise<number> {
    try {
      const elements = await this.driver!.findElements(
        By.css('article div div div span')
      );
      
      for (const element of elements) {
        const text = await element.getText();
        if (text.includes('post')) {
          return this.parseCount(text.split(' ')[0]);
        }
      }
      return 0;
    } catch {
      return 0;
    }
  }

  private async getBio(): Promise<string> {
    try {
      // Birden fazla selector dene
      const selectors = [
        'article header section div div span', // Yeni Instagram yapısı
        'article header section div span', // Alternatif
        'header section div div span', // Daha genel
        'article header section div', // Eski selector
        '[data-testid="user-bio"]', // Test ID varsa
        'div[dir="auto"] span' // Bio text için
      ];

      for (const selector of selectors) {
        try {
          const elements = await this.driver!.findElements(By.css(selector));
          
          for (const element of elements) {
            const text = await element.getText();
            // Bio olabilecek text'i kontrol et (boş değil ve çok kısa değil)
            if (text && text.length > 5 && !text.includes('posts') && !text.includes('followers')) {
              console.log(`Bio found with selector: ${selector}, text: ${text.substring(0, 100)}...`);
              return text;
            }
          }
        } catch (e) {
          // Bu selector çalışmadı, devam et
          continue;
        }
      }

      console.log('Bio not found with any selector');
      return '';
    } catch (error) {
      console.error('Error getting bio:', error);
      return '';
    }
  }

  private async isVerified(): Promise<boolean> {
    try {
      await this.driver!.findElement(By.css('[aria-label="Verified"]'));
      return true;
    } catch {
      return false;
    }
  }

  private async isPrivate(): Promise<boolean> {
    try {
      const text = await this.driver!.getPageSource();
      return text.includes('This account is private');
    } catch {
      return false;
    }
  }

  private async getRecentPosts(): Promise<InstagramPost[]> {
    try {
      // Son 6 postu al
      const postElements = await this.driver!.findElements(
        By.css('article div div div div a')
      );

      const posts: InstagramPost[] = [];
      
      for (let i = 0; i < Math.min(6, postElements.length); i++) {
        try {
          const post = postElements[i];
          const href = await post.getAttribute('href');
          
          // Post detaylarını al (basit versiyon)
          const img = await post.findElement(By.css('img'));
          const imageUrl = await img.getAttribute('src');
          
          posts.push({
            imageUrl,
            caption: '', // Detay sayfasından alınabilir
            likes: 0,
            comments: 0,
            timestamp: '',
            hashtags: []
          });
        } catch (error) {
          console.error('Error getting post:', error);
        }
      }

      return posts;
    } catch {
      return [];
    }
  }

  private parseCount(text: string): number {
    if (!text) return 0;
    
    const cleanText = text.replace(/,/g, '').toLowerCase();
    
    if (cleanText.includes('k')) {
      return Math.floor(parseFloat(cleanText) * 1000);
    } else if (cleanText.includes('m')) {
      return Math.floor(parseFloat(cleanText) * 1000000);
    } else {
      return parseInt(cleanText) || 0;
    }
  }

  async close() {
    if (this.driver) {
      await this.driver.quit();
      this.driver = null;
    }
  }
}

// Kullanım örneği
export async function scrapeCompetitors(usernames: string[]): Promise<InstagramProfile[]> {
  const scraper = new InstagramScraper();
  await scraper.initialize();

  const profiles: InstagramProfile[] = [];
  
  for (const username of usernames) {
    const profile = await scraper.scrapeProfile(username);
    if (profile) {
      profiles.push(profile);
    }
    
    // Rate limiting - 2 saniye bekle
    await new Promise(resolve => setTimeout(resolve, 2000));
  }

  await scraper.close();
  return profiles;
}